plugin/src/main/java/com/google/errorprone/xplat/checker/LazyInitBan.java

```diff
@@ -0,0 +1,409 @@
+    summary = "Lazy Init is an error prone pattern on iOS. Please do it by the book.",
+    explanation =
+        "TBD",
+    severity = ERROR)
+public class LazyInitBan extends BugChecker implements MethodTreeMatcher {
+
+  private static final Matcher<MethodTree> METHOD_MATCHER =
+      Matchers.allOf(
+          Matchers.not(Matchers.methodIsConstructor())
+      );
+
+  /**
+   * Used by IfTreeNonSync as a return value to get needed information across.
+   */
+  private enum IfTreeReturn {
+    NO_MATCH, MATCH, WRONG_ASSIGNMENT_ORDER
+  }
+
+  /*
+   * Holds needed information for error messages when WRONG_ASSIGNMENT_ORDER is found.
+   */
+  private static class ExpressionBox {
+
+    public ExpressionTree tree;
+    public Name firstVar;
+    public Name secondVar;
+
+    ExpressionBox() {
+
+    }
+
+    public void editBox(ExpressionTree tree, Name firstVar, Name secondVar) {
+      this.tree = tree;
+      this.firstVar = firstVar;
+      this.secondVar = secondVar;
+    }
+  }
+
+  /**
+   * Checks the given field to see if it is "correct" for the lazy init pattern.
+   *
+   * @param field      symbol of field
+   * @param methodSync whether the method is synchronized
+   * @return true if the check passes
+   */
+  private boolean checkFieldSymbolSync(Symbol field, boolean methodSync) {
+    boolean annotation = field.getAnnotationsByType(LazyInit.class).length > 0;
+    boolean is_field = field.getKind() == ElementKind.FIELD;
+    boolean is_final = field.getModifiers().contains(Modifier.FINAL);
+    boolean is_vol = field.getModifiers().contains(Modifier.VOLATILE);
+    return is_field && (((is_vol || annotation) && methodSync) || is_final);
+  }
+
+  /**
+   * Checks if the given IfTree matches one of the invalid patterns.
+   *
+   * @param ifTree      if statement tree
+   * @param foundIdents the set of found identifiers
+   * @param methodSync  whether the method is synchronized
+   * @return whether a pattern matches or not
+   */
+  private boolean ifTreeSync(IfTree ifTree, Set<Name> foundIdents, boolean methodSync) {
+    ExpressionTree exp = ifTree.getCondition();
+
+    if (exp.getKind() != Kind.PARENTHESIZED) {
+      return false;
+    }
+
+    exp = ((ParenthesizedTree) exp).getExpression();
+
+    if (exp.getKind() != Kind.EQUAL_TO) {
+      return false;
+    }
+
+    ExpressionTree left = ((BinaryTree) exp).getLeftOperand();
+
+    ExpressionTree right = ((BinaryTree) exp).getRightOperand();
+
+    if (right.getKind() != Kind.NULL_LITERAL || left.getKind() != Kind.IDENTIFIER) {
+      return false;
+    }
+
+    if (ASTHelpers.getSymbol(left).isLocal()) {
+      return false;
+    }
+
+    Symbol fieldSymbol = ASTHelpers.getSymbol(left).enclClass().members().findFirst(
+        (com.sun.tools.javac.util.Name) ((IdentifierTree) left).getName());
+
+    if (fieldSymbol == null || checkFieldSymbolSync(fieldSymbol, methodSync)) {
+      return false;
+    }
+
+    StatementTree tree = ifTree.getThenStatement();
```

Comment:
What happens if one writes `if (local == null) field = local = computeField();` without the block braces?

Reply:
It did not function properly. I have updated the code to fix this.